### DISCUSSION TITLE
Similar to filter-adjustment table: digi-volume table

### DISCUSSION THREAD

**Discussion by Toddler-Boy** [libsidplayfp/libsidplayfp#100] [Ideas]
Similar to creating an author table to adjust 6581 filter settings, we could also adjust digi-volume for the 6581 so the DC offset can be changed. Some tunes have very loud digis, others are average, and some are very quiet.

**Comment by drfiemost** [comment]
The DC offset is not exposed as a tunable yet, need to find a sensible range and it would affect the filter too.
It could be an idea for the future though.

**Comment by Toddler-Boy** [comment]
How can I change the amount of distortion? What internal variable/constant controls that? It seems very strong for most tunes.

Unfortunately, the current reSIDfp 6581-filter settings are tuned so much towards a "Martin Galway Wizball sounds correct" state of mind, that everything else seems like an afterthought.

In an ideal world, we should be able to independently adjust the filter offset, filter range, filter distortion, and digi-volume. Then, we could create chip profiles and assign them to various composers so their tunes sound good/as intended, not merely recognizable because the melody is cool.

**Comment by drfiemost** [comment]
It's not that easy, there are a few parameter to play with. Check out the [filter config](https://github.com/libsidplayfp/libsidplayfp/blob/f6fdbc9d0f5012eff2c6b01ffb93c40610177f68/src/builders/residfp-builder/residfp/FilterModelConfig6581.cpp#L118-L123), especially the uCox value.
Also keep in mind that the [opamp voltage table](https://github.com/libsidplayfp/libsidplayfp/blob/f6fdbc9d0f5012eff2c6b01ffb93c40610177f68/src/builders/residfp-builder/residfp/FilterModelConfig6581.cpp#L59) has been sampled on a 6581R4AR and may not be representative of other revisions.
Then there is also the [voice offset](https://github.com/libsidplayfp/libsidplayfp/blob/f6fdbc9d0f5012eff2c6b01ffb93c40610177f68/src/builders/residfp-builder/residfp/SID.cpp#L109).

**Comment by Toddler-Boy** [comment]
Can you provide us with more details on what these values do? What happens if I lower or raise the ```uCox``` value? What does it represent? What's a sensible range?

These are all magic numbers with no documentation.

Everything around 5v should (according to specs) be precisely 5v. Same with 12v. Measuring one specific chip in one specific C64 with a particular power supply and applying it to everything seems rather arbitrary.

Is my only way to figure this out is to change those numbers, recompile, and listen to differences?

Is there any help you can provide from experience?

Let's say I would like to lower the overall filter distortion. Is that something the opamp table is responsible for? Or rather, the voltages/capacities in the filter-config?

**Reply by drfiemost** [reply]
> Can you provide us with more details on what these values do? What happens if I lower or raise the uCox value? What does it represent? What's a sensible range?
> 
> These are all magic numbers with no documentation.

uCox is a MOSFET parameter and depends on the chip production tech used. You can find some info on the net but nothing specific to the '80s technology so it's mostly guessing and tuning.

> Everything around 5v should (according to specs) be precisely 5v. Same with 12v. Measuring one specific chip in one specific C64 with a particular power supply and applying it to everything seems rather arbitrary.

The C64 power bricks were quite bad so it's likely the voltages were off spec but yes, as with other parameters the choice is arbitrary. We could make these tunable in a small range around the spec but I'm not sure it's worth the effort.

I'm afraid you have to play with the values yourself (and hope it won't crash). Wilfred have noted some of his results in #45 and you can read A. Lankila's notes on 6581 distortion [here](https://web.archive.org/web/20221007232553/https://bel.fi/alankila/c64-sw/index-cpp.html)

**Comment by Toddler-Boy** [comment]
Thanks. I've saved Lankila's page as a local website (with images, etc.) so as not to rely on web archives.

I guess I will have to play around with these parameters and see what happens.

**Comment by Toddler-Boy** [comment]
I think the uCox parameter is the single most important parameter in the filter config. Using the current filter adjustment is just wrong. It also affects the balance between the voices, shifts the distortion into unpleasant territory, etc., whereas adjusting the uCox has none of these side effects while greatly affecting the cutoff position.

I've multiplied the uCox with 0.2, and I dare to say that ```/MUSICIANS/D/Dunn_David/Jump_Jet.sid``` sounds almost perfect now.

I've done similar tests with other tunes (different uCox multipliers while leaving the filterOffset at its default 0.5) and the results are much better than the previous method.

Unfortunately, that also means that caching the entire filter-config is no longer possible, as the ```vcr_n_Ids_term``` table has to be calculated every time the uCox changes, but that's only 1/6 of the entire cache, so it should be easy to split these six into individual caches and invalidate the last one if the uCox value changes.

**Comment by drfiemost** [comment]
I agree the uCox is a good candidate for becoming a tunable, also for the 8580. Getting the value wrong can cause a crash though so we need a safe range to work with.

**Comment by Toddler-Boy** [comment]
I've added a function to tweak it for each composer (as in our previous discussion), and it's a multiplier between 0.05 and 2.0. That value gets multiplied by 20e-6, and there are no crashes. You may have encountered a division by zero or something like that.

```uCox = std::clamp ( adjustment, 0.05, 2.0 ) * 20e-6;```

I'm currently adjusting the range (that's what this parameter seems to represent), and the results are much better than adjusting the DAC offset.

Every single tune I've tried so far (a couple hundred) sounds much better than before. It's insane.

Let me know if you would like some examples (before and after) and which tunes you would like. The Martin Galway stuff is unnecessary, as their current filter settings have been tweaked for "Wizball".

**Comment by drfiemost** [comment]
Cool! I'l add it as a TODO item for the next version.

**Comment by drfiemost** [comment]
Just committed a fix for the 6581 filter: https://github.com/libsidplayfp/libsidplayfp/commit/0c4f6858700b33e3e2856d4cbe2cfeb0a688b3f1
It shouldn't alter the sound color that much, let me know your results.
Also while testing I recommend enabling assertions to check that variables don't go outside the allowed range.

**Comment by Toddler-Boy** [comment]
I've implemented the same changes to my version, and it works really well. As you said, not much change to the sound color.

In my debug builds, assertions are always enabled. No negative values here.

BTW, the members min/max in ```src/builders/residfp-builder/residfp/Integrator6581.h``` are terribly named and are not even used. I would recommend removing them.

**Comment by drfiemost** [comment]
> BTW, the members min/max in `src/builders/residfp-builder/residfp/Integrator6581.h` are terribly named and are not even used. I would recommend removing them.

Whoops, that was just some debug stuff that accidentally went into the commit, now removed, thanks!

**Comment by Toddler-Boy** [comment]
New problem: if we recalculate the ```vcr_n_Ids_term``` LUT for filter-range changes, the code instantly becomes non-thread safe again. Because of the caching, all threads would share the same settings, so different threads changing the filter range would affect all threads. It wouldn't lead to a crash, but rather to mid-rendering filter changes.

To work around that, while also avoiding recreating these tables every time a new tune gets loaded, or SID gets added or removed, I've made the FilterModelConfig something external to the Filters, and their constructors now expect a reference to the correct FilterModelConfig. This way, all LUTs need to be created only once per thread and are not shared between threads.

Technically, the first five LUTs could be shared, but on my machine, with the LUT creation being parallelized via std::threads, the whole thing only takes 50-60ms, and then my player object has the FilterModelConfig, so it never gets re-created, I decided it's not worth it for my project.

Addendum: I must have broken something and reverted to the global version. Then I had an idea. Now I use ```thread_local``` instead of a static instance, removed the mutex, but otherwise left the same logic. Had to make the constructor and destructor public, but that's a small price to pay. Now it's thread-safe by default, no mutexes needed and everything works as intended. FilterModelConfigs only get created once per thread.
